### DISCUSSION TITLE
Is there a way to propagate redirects through sequenced Handle hooks?

### DISCUSSION THREAD

**Discussion by MikeMcDonald83** [sveltejs/kit#10754] [Q&A] [answered]
I have my server-session and authentication handles in 3 different functions. They are sequenced with the sequence helper. As simplified examples:
- Check for session ID and load redis data into locals.
```
const load = async ({event, resolve}) => {
  const sid = event.cookies.get("myauthcookie");
  event.locals = redisGetAndParse(sid);

  return await resolve(event);
}
```
- Do auth stuff, including adding or removing stuff from locals.
```
const auth = async ({event, resolve}) => {
  if (route === protected && user !== loggedIn) {
    event.locals.nonce = getCuid();
    event.locals.state = getCuid();
    const qParams = new URLSearchParams(...);

    REDIRECT_SOMEHOW(authURL + "?" + qParams);
  } else {
    return await resolve(event);
  }
}
```
- Save locals back into redis.
```
const save = async ({event, resolve}) => {
  const sid = event.cookies.get("myauthcookie");
  redisStringifyAndSet(sid, event.locals);

  return await resolve(event);
}
```

The problem is that my auth Handle function needs to redirect sometimes. I can't seem to make it to the third function if I redirect in any way. It is preventing me from storing my state/nonce/returnTo session variables before the request ends. In the auth Handle, I've tried `throw redirect(300, "http://mykeycloak")`. I've tried `return new Response(null, {status: 300, headers: {location: "http://mykeycloak"}}`. I've tried `event.url = "http://mykeycloak"`.

I could probably pass a message to the third Handle to tell it to redirect, but I'm trying to keep all the session and auth logic separate while making use of existing features. I could probably pass a `save()` function from my session class to my auth class. The same considerations would apply though.

Can someone please point me in the right direction? Am I abusing sequence? Am I missing something obvious? Thanks.

**Comment by david-plugge** [comment, marked as answer]
You probably want to rethink about how you implemented your `load` and `save` handle hooks.
To me it looks like you want to populate `locals` before sveltekit does it´s stuff and _save_ the state to redis when its done.
If that is the case, you can (and should) combine the logic into a single handle hook:

```ts
const load = async ({ resolve, event }) => {
  const sid = event.cookies.get("myauthcookie");

  // load the state
  event.locals = redisGetAndParse(sid);

  // sveltekit stuff
  const response = await resolve(event);

  // response may be a redirect but we don´t really care
  // we just want to do some stuff after the request is created

  // save the state
  redisStringifyAndSet(sid, event.locals);

  return response;
}
```

**Reply by MikeMcDonald83** [reply]
Works great. To be honest, I stumbled on the `sequence` helper in the docs and tried to find a problem for the solution. After prodding around some more, I realized that I was trying to figure out how a response from `resolve(event)` invokes the next Handle while a redirect response does not. In those terms, it seems likely that a redirect response it's missing something from the original request event (custom headers or something). Anyways, thanks for your time.

**Comment by machak** [comment]
You can try 
`return Response.redirect(`${event.url.origin}/some/url`, 302);`
